### PR TITLE
Adding the first version of a misk backfila client.

### PR DESCRIPTION
**Pull request by mpawliszyn**
Adding HibernateBackfill with testing.

### CODE REVIEW

**Comment by mpawliszyn** [comment]
I need https://github.com/cashapp/misk/pull/1298 merged and then I'll need to update the dependencies. It builds locally.

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/EmbeddedBackfila.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    check(request.connector_type == Connectors.HTTP) {"Misk client only supports HTTP."}
+    serviceData = request
+    return Calls.response(ConfigureServiceResponse())
+  }
+
+  fun <Type: Backfill> createDryRun(
+    backfill: KClass<Type>,
+    parameters: Map<String, ByteString> = mapOf(),
+    rangeStart: String? = null,
+    rangeEnd: String? = null): EmbeddedBackfillRun<Type> {
```
Comment (shellderp) [inline client-misk/src/main/kotlin/app/cash/backfila/client/misk/EmbeddedBackfila.kt]:
```suggestion
    rangeEnd: String? = null
  ): EmbeddedBackfillRun<Type> {
```

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/EmbeddedBackfila.kt [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    return createBackfill(backfill, true, parameters, rangeStart, rangeEnd)
+  }
+
+  fun <Type: Backfill> createWetRun(
+    backfillType: KClass<Type>,
+    parameters: Map<String, ByteString> = mapOf(),
+    rangeStart: String? = null,
+    rangeEnd: String? = null): EmbeddedBackfillRun<Type> {
```
Comment (shellderp) [inline client-misk/src/main/kotlin/app/cash/backfila/client/misk/EmbeddedBackfila.kt]:
```suggestion
    rangeEnd: String? = null
  ): EmbeddedBackfillRun<Type> {
```

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/EmbeddedBackfila.kt [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    return createBackfill(backfillType, false, parameters, rangeStart, rangeEnd)
+  }
+
+  // TODO(mikepaw) Maybe we just annotate a data class with @Parameters or perhaps a data class that has a fromByteStringMap method and then we pass in the data object?
+  private fun <Type: Backfill> createBackfill(
+    backfillType: KClass<Type>,
+    dryRun: Boolean, parameters: Map<String, ByteString>,
+    rangeStart: String?,
+    rangeEnd: String?): EmbeddedBackfillRun<Type> {
```
Comment (shellderp) [inline client-misk/src/main/kotlin/app/cash/backfila/client/misk/EmbeddedBackfila.kt]:
```suggestion
    rangeEnd: String?
  ): EmbeddedBackfillRun<Type> {
```

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/EmbeddedBackfillRun.kt [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+import app.cash.backfila.protos.clientservice.GetNextBatchRangeResponse
+import app.cash.backfila.protos.clientservice.KeyRange
+import app.cash.backfila.protos.clientservice.PrepareBackfillRequest
+import app.cash.backfila.protos.clientservice.PrepareBackfillResponse
+import app.cash.backfila.protos.clientservice.RunBatchRequest
+import okio.ByteString
+import okio.ByteString.Companion.encodeUtf8
+import java.util.ArrayDeque
+
+//TODO(mikepaw) do I need to do anything to make this more explicitly not thread safe?
```
Comment (shellderp) [inline client-misk/src/main/kotlin/app/cash/backfila/client/misk/EmbeddedBackfillRun.kt]: is this class only used for tests? then no need for thread safe. I wonder if we want a `client-misk-test` module so you don't always pull in the testing code

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/base/HibernateBackfill.kt [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+import okio.ByteString
+import okio.ByteString.Companion.encodeUtf8
+import java.util.ArrayList
+import java.util.concurrent.TimeUnit
+import javax.inject.Inject
+import javax.persistence.Table
+import kotlin.reflect.KClass
+
+/**
+ * Generic backfill that runs over Hibernate 4 entities.
```
Comment (shellderp) [inline client-misk/src/main/kotlin/app/cash/backfila/client/misk/base/HibernateBackfill.kt]: are you planning to abstract some of this to share with square internal code?

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaClientModule.kt [hunk 6]
```diff
@@ -35,22 +37,18 @@ annotation class ForBackfila
 
 class BackfilaClientModule(
   private val config: BackfilaClientConfig,
-  private val backfills: List<Class<out Backfill>>
+  private val backfills: List<KClass<out Backfill>>
```
Comment (shellderp) [inline client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaClientModule.kt]: what's the reason to convert to KClass? Curious

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/EmbeddedBackfillRun.kt [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+        .backfill_name(backfill.javaClass.toString())
+        .range(KeyRange(rangeStart?.encodeUtf8(), rangeEnd?.encodeUtf8()))
+        .parameters(parameters)
+        .dry_run(dryRun)
+        .build())
+    prepareBackfillResponse = response
+    precomputeProgress = response.instances.map { Pair(
+        it.instance_name,
+        InstanceCursor(it.instance_name,
+            it.backfill_range)) }.toMap()
```
Comment (shellderp) [inline client-misk/src/main/kotlin/app/cash/backfila/client/misk/EmbeddedBackfillRun.kt]: nit `}` on new line

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/EmbeddedBackfillRun.kt [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+    scanProgress = response.instances.map { Pair(
+        it.instance_name,
+        InstanceCursor(it.instance_name,
+            it.backfill_range)) }.toMap()
+  }
+
+  // Precomputing Methods
+  fun precomputeScan(): GetNextBatchRangeResponse {
+    val notDone = precomputeProgress.filter { !it.value.done }
+    check(notDone.isNotEmpty()) {"Nothing left to precompute"}
```
Comment (shellderp) [inline client-misk/src/main/kotlin/app/cash/backfila/client/misk/EmbeddedBackfillRun.kt]: nit: add spaces inside `{}`
```suggestion
    check(notDone.isNotEmpty()) { "Nothing left to precompute" }
```

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/base/BackfillConfig.kt [hunk 9]
```diff
@@ -0,0 +1,9 @@
+package app.cash.backfila.client.misk.base
+
+import app.cash.backfila.protos.service.Parameter
+import okio.ByteString
+
+data class BackfillConfig(
+  val parameters: Map<String,ByteString>,
```
Comment (shellderp) [inline client-misk/src/main/kotlin/app/cash/backfila/client/misk/base/BackfillConfig.kt]:
```suggestion
  val parameters: Map<String, ByteString>,
```

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/base/HibernateBackfill.kt [hunk 10]
```diff
@@ -0,0 +1,411 @@
...
+   * Separate from primaryKeyName() as the casing is usually different.
+   */
+  protected open fun primaryKeyHibernateName(): String = "id"
+
+  /**
+   * Criteria that filters which records are selected to backfill from the table.
+   *
+   * This must return a new instance of Query in every invocation.
+   */
+  protected abstract fun backfillCriteria(config: BackfillConfig): Query<E>
```
Comment (shellderp) [inline client-misk/src/main/kotlin/app/cash/backfila/client/misk/base/HibernateBackfill.kt]: neat!

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/base/HibernateBackfill.kt [hunk 11]
```diff
@@ -0,0 +1,411 @@
...
+          .build()
+    }
+
+    val stopwatch = Stopwatch.createStarted()
+    val batchGenerator = BatchGenerator(request)
+
+    val batches = ArrayList<Batch>()
+    while (batches.size < request.compute_count_limit) {
+      val batch = batchGenerator.next()
+          ?: // No more batches can be computed.
```
Comment (shellderp) [inline client-misk/src/main/kotlin/app/cash/backfila/client/misk/base/HibernateBackfill.kt]: weird formatting

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/base/HibernateBackfill.kt [hunk 12]
```diff
@@ -0,0 +1,411 @@
...
+    private var boundingMax: Pkey? = null
+
+    init {
+      instanceName = request.instance_name
+      batchSize = request.batch_size
+      scanSize = request.scan_size
+      backfillRange = request.backfill_range
+      previousEndKey = request.previous_end_key
+      config = BackfillConfig(request.parameters, request.dry_run)
+      precomputing = java.lang.Boolean.TRUE == request.precomputing
```
Comment (shellderp) [inline client-misk/src/main/kotlin/app/cash/backfila/client/misk/base/HibernateBackfill.kt]:
```suggestion
      precomputing = request.precomputing == true
```

File: client-misk/src/test/kotlin/app/cash/backfila/client/misk/base/UnshardedHibernateBackfillTest.kt [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+class UnshardedHibernateBackfillTest {
+  @Suppress("unused")
+  @MiskTestModule
+  val module: Module = ClientMiskTestingModule(false, listOf(TestBackfill::class))
+
+  @Inject @ClientMiskService lateinit var transacter: Transacter
+  @Inject lateinit var embeddedBackfila: EmbeddedBackfila
+
+  class TestBackfill @Inject constructor(
+    entityType: TypeLiteral<DbMenu>,
```
Comment (shellderp) [inline client-misk/src/test/kotlin/app/cash/backfila/client/misk/base/UnshardedHibernateBackfillTest.kt]: is there a way to not have to provide these in every backfill?

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaClientModule.kt (hunk 6 above)
Comment (mpawliszyn) [reply client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaClientModule.kt]: Query uses KClass further down. I figured in Kotlin we might want to use KClass over Class?

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/EmbeddedBackfillRun.kt (hunk 4 above)
Comment (mpawliszyn) [reply client-misk/src/main/kotlin/app/cash/backfila/client/misk/EmbeddedBackfillRun.kt]: Yeah I guess I could extract it out. Might be a good idea.

File: client-misk/src/test/kotlin/app/cash/backfila/client/misk/base/UnshardedHibernateBackfillTest.kt (hunk 13 above)
Comment (mpawliszyn) [reply client-misk/src/test/kotlin/app/cash/backfila/client/misk/base/UnshardedHibernateBackfillTest.kt]: Yeah I'm thinking about it. Not sure if there is a better way to approach this problem. Might be one with Kotlin. I'll investigate this later though.

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/base/HibernateBackfill.kt (hunk 5 above)
Comment (mpawliszyn) [reply client-misk/src/main/kotlin/app/cash/backfila/client/misk/base/HibernateBackfill.kt]: We will see. Right now I'm focused on only getting the kotlin one right. Then I'll see what can be ported over.

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaClientModule.kt (hunk 6 above)
Comment (adrw) [reply client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaClientModule.kt]: For a Kotlin client I'd also lean with `KClass`

File: client-misk/src/test/kotlin/app/cash/backfila/client/misk/base/UnshardedHibernateBackfillTest.kt [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+      if (config.dryRun) {
+        idsRanDry.addAll(pkeys)
+      } else {
+        idsRanWet.addAll(pkeys)
+      }
+    }
+  }
+
+  @Test fun emptyTable() {
+    val run = embeddedBackfila.createDryRun(TestBackfill::class).configureForTest()
```
Comment (adrw) [inline client-misk/src/test/kotlin/app/cash/backfila/client/misk/base/UnshardedHibernateBackfillTest.kt]: Using a inline reified function would be pretty cool 👍

```
// definition
inline fun EmbeddedBackfila.createDryRun<reified T>() = createDryRun(T::class)

// usage
embeddedBackfila.createDryRun<TestBackfill>()
```

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaClientModule.kt (hunk 6 above)
Comment (mpawliszyn) [reply client-misk/src/main/kotlin/app/cash/backfila/client/misk/BackfilaClientModule.kt]: They are pretty much interchangeable. Jesse didn't seem to care but agreed with my reasoning.

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/EmbeddedBackfillRun.kt (hunk 4 above)
Comment (mpawliszyn) [reply client-misk/src/main/kotlin/app/cash/backfila/client/misk/EmbeddedBackfillRun.kt]: The comment has been updated

File: client-misk/src/test/kotlin/app/cash/backfila/client/misk/base/UnshardedHibernateBackfillTest.kt (hunk 14 above)
Comment (mpawliszyn) [reply client-misk/src/test/kotlin/app/cash/backfila/client/misk/base/UnshardedHibernateBackfillTest.kt]: Yeah I ended up doing this.

File: client-misk/src/test/kotlin/app/cash/backfila/client/misk/base/UnshardedHibernateBackfillTest.kt (hunk 13 above)
Comment (mpawliszyn) [reply client-misk/src/test/kotlin/app/cash/backfila/client/misk/base/UnshardedHibernateBackfillTest.kt]: Yeah the new code makes this go away!

**Comment by mpawliszyn** [comment]
I'm going to merge this reworked version in. I can make changes later for any additional comments people have post the latest refactor. cc: @adrw @shellderp

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/EmbeddedBackfillRun.kt [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+  /** Prepares a backfill for a run. */
+  init {
+    prepareBackfillResponse = operator.prepareBackfill(PrepareBackfillRequest.Builder()
+        .backfill_name(operator.name())
+        .range(KeyRange(rangeStart?.encodeUtf8(), rangeEnd?.encodeUtf8()))
+        .parameters(parameters)
+        .dry_run(dryRun)
+        .build())
+    precomputeProgress = prepareBackfillResponse.instances.associate {
+      it.instance_name to MutableInstanceCursor(
```
Comment (shellderp) [inline client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/EmbeddedBackfillRun.kt]: bring the next line up

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/EmbeddedBackfillRun.kt [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+          it.instance_name, it.backfill_range)
+    }
+    scanProgress = prepareBackfillResponse.instances.associate {
+      it.instance_name to MutableInstanceCursor(
```
Comment (shellderp) [inline client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/EmbeddedBackfillRun.kt]: same

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/EmbeddedBackfillRun.kt [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+      response.batches.onEach {
+        precomputeMatchingCount += it.matching_record_count
+        precomputeScannedCount += it.scanned_record_count
+      }
+    }
+    return response
+  }
+
+  override fun precomputeRemaining() {
+    do {
```
Comment (shellderp) [inline client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/EmbeddedBackfillRun.kt]: aww yeah do while

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/EmbeddedBackfillRun.kt [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+            .dry_run(dryRun)
+            .build())
+    when (response.batches.isEmpty()) {
+      true -> cursor.done = true
+      false -> cursor.previousEndKey = response.batches.map { it.batch_range.end }.max()
+    }
+    batchesToRun.addAll(response.batches
+        .filterNot { it.matching_record_count == 0L } // Remove batches that have no matching records.
+        .map {
+          BatchSnapshot(instanceName,
```
Comment (shellderp) [inline client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/EmbeddedBackfillRun.kt]: putting every arg on its own line is more kotliny

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/EmbeddedBackfillRun.kt [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+    return response
+  }
+
+  override fun singleScan(): GetNextBatchRangeResponse {
+    val cursor = scanProgress.values.first { !it.done }
+    return instanceScan(cursor.instanceName)
+  }
+
+  override fun scanRemaining() {
+    do singleScan()
```
Comment (shellderp) [inline client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/EmbeddedBackfillRun.kt]: lack of {} on a do while looks really weird

File: client-misk/src/test/kotlin/app/cash/backfila/client/misk/ClientMiskTestingModule.kt [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+import misk.logging.LogCollectorModule
+import kotlin.reflect.KClass
+
+internal class ClientMiskTestingModule(
+  val useVitess: Boolean,
+  val backfillClasses: List<KClass<out Backfill<*, *>>>
+) : KAbstractModule() {
+  override fun configure() {
+    val dataSourceConfig =  when {
+      useVitess ->  DataSourceConfig(
```
Comment (shellderp) [inline client-misk/src/test/kotlin/app/cash/backfila/client/misk/ClientMiskTestingModule.kt]: formatting

File: client-misk/src/test/kotlin/app/cash/backfila/client/misk/base/UnshardedHibernateBackfillTest.kt [hunk 21]
```diff
@@ -0,0 +1,282 @@
...
+@MiskTest(startService = true)
+class UnshardedHibernateBackfillTest {
+  @Suppress("unused")
+  @MiskTestModule
+  val module: Module = ClientMiskTestingModule(false, listOf(TestBackfill::class))
+
+  @Inject @ClientMiskService lateinit var transacter: Transacter
+  @Inject lateinit var backfila: Backfila
+
+  // (pkeySqlAdapter, queryFactory, transacter)
```
Comment (shellderp) [inline client-misk/src/test/kotlin/app/cash/backfila/client/misk/base/UnshardedHibernateBackfillTest.kt]: remove?

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/BackfilaWebApiActions.kt [hunk 22]
```diff
@@ -88,14 +47,16 @@ internal class GetNextBatchRangeAction @Inject constructor(
...
 
-    val backfill = backfilaClassResolver.getBackfill(request.backfill_name, request.backfill_id)
-    val nextBatchRange = backfill.getNextBatchRange(request)
-    logger.info {
-      "Next batches computed for backfill " +
-          "`${request.backfill_name}::${request.instance_name}::${request.backfill_id}`. " +
-          "${nextBatchRange.batches}"
-    }
-    return nextBatchRange
+    val operator = operatorFactory.create(request.backfill_name, request.backfill_id)
```
[the remark this answers is not in the document]
Comment (mpawliszyn) [reply client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/BackfilaWebApiActions.kt]: I think we need that other example and then we could extract a base and hibernate operator here leaving the other to implement the other side.

The operatorFactory should have enough to decide the right type with the backfill_name though no?

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/BackfillOperator.kt [hunk 23]
```diff
@@ -0,0 +1,408 @@
...
+        | ORDER BY ${pkeyName}
+        | LIMIT ${scanSize}) s
+        """.trimMargin()
+    val max = session.hibernateSession.createNativeQuery(sql).uniqueResult()
+
+    @Suppress("UNCHECKED_CAST") // Return type from the query should always match.
+    return max as Pkey? // TODO(mikepaw,jwilson) running through the tests I think we are always getting a Pkey. I think this remains true all the time.
+  }
+
+  private inner class BatchGenerator internal constructor(request: GetNextBatchRangeRequest) {
```
[the remark this answers is not in the document]
Comment (mpawliszyn) [reply client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/BackfillOperator.kt]: It uses some methods that are on the Backfill Operator.

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/EmbeddedBackfillRun.kt (hunk 17 above)
Comment (mpawliszyn) [reply client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/EmbeddedBackfillRun.kt]: I can put that all on one line too. Very easy to understand. Kotlin is great.

Comment (mpawliszyn) [reply client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/EmbeddedBackfillRun.kt]: The autoformatter doesn't like it though. So I'm leaving it alone.

File: client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/EmbeddedBackfillRun.kt (hunk 19 above)
Comment (mpawliszyn) [reply client-misk/src/main/kotlin/app/cash/backfila/client/misk/internal/EmbeddedBackfillRun.kt]: Ok. has brackets now.

File: client-misk/src/test/kotlin/app/cash/backfila/client/misk/ClientMiskTestingModule.kt (hunk 20 above)
Comment (mpawliszyn) [reply client-misk/src/test/kotlin/app/cash/backfila/client/misk/ClientMiskTestingModule.kt]: Not 100% sure I know what you are looking for. But I autoformatted.

File: client-misk/src/test/kotlin/app/cash/backfila/client/misk/base/UnshardedHibernateBackfillTest.kt (hunk 21 above)
Comment (mpawliszyn) [reply client-misk/src/test/kotlin/app/cash/backfila/client/misk/base/UnshardedHibernateBackfillTest.kt]: yeah I had already done this.

**Comment by mpawliszyn** [comment]
It will be in the next PR.

**Pull request merged**